### PR TITLE
Update the cluster API per the latest decisions

### PR DESCRIPTION
**Pull request by perotinus**
This is intended to get the API very close to an alpha state.

cf #20 

cc @font @madhusudancs @pmorie

### CODE REVIEW

**Comment by madhusudancs** [comment]
I am really confused. Copying Endpoints object seems incorrect. For example, what purpose would `EndpointPort.Protocol` serve? If anything, this should mirror what's in federation now. I can't see all the fields in Kubernetes core api/v1.Endpoint being appropriate here.

**Comment by perotinus** [comment]
@madhusudancs If the concern is that `EndpontPort.Protocol` is not useful, we can remove it.

@quinton-hoole, what was your reasoning behind suggesting using`api.Endpoints` rather than mimicking what Federation currently has?

**Comment by perotinus** [comment]
@madhusudancs, are you OK with this going in like this for now, with the linked issue preserving a question about using the Endpoints object?

**Comment by madhusudancs** [comment]
I am Ok with this change going in without the Endpoints change. It's more work to revert it than separating the parts under question into a separate commit.

**Comment by perotinus** [comment]
@font @pmorie @madhusudancs Any other issues? I'd like to submit this; we can iterate further if necessary.

**Comment by madhusudancs** [comment]
@perotinus I don't see any change to the PR since I last looked. Did you push a new commit? I am still on removing the endpoints change in this PR until it is settled. No other issues. If you remove that change along with the dependent types, I am fine lgtm'ing this PR.

**Comment by tengqm** [comment]
I tend to agree with @madhusudancs that introducing a new 'Endpoint' concept may cause confusion to people. We have to explain to users that this Endpoint is not the one they see from kubernetes ... If possible, we may want to avoid the Cartesian product computation ... we may want to support just HTTP/HTTPS (instead of UDP/TCP)... After all, we are talking about a Kubernetes API server instead of a generic service endpoint to be supported by Kubernetes.

**Comment by ericchiang** [comment]
We use a load balancer behind a DNS address for our HA API servers. All clients use the load balancer address. We never refer to the API servers through a bare IP address.

Why did the API switch to only supporting IPs?

**Comment by perotinus** [comment]
@ericchiang The API as it is still supports hostnames: https://github.com/kubernetes/cluster-registry/pull/31/files#diff-7d367230bef0cc50523964c6d3dccdcbR206 

I admit the naming is confusing.

Given the amount of pushback, I'm going to change this to something like what Federation does today. I believed that this had been addressed in the API doc, since there was a lack of objection to using `api.Endpoints`, but we'll have to have some more explicit discussion in a follow-up.

cc @quinton-hoole, who suggested the `api.Endpoints` type

**Comment by perotinus** [comment]
@ericchiang @madhusudancs PTAL.

**Comment by pmorie** [comment]
Soery i missed your earlier tag - gh notifications are pretty useless for
me. Will TALat this tonight.

File: pkg/apis/clusterregistry/types.go [hunk 1]
```diff
@@ -32,43 +32,64 @@ type Cluster struct {
...
+}
+
+// ClusterStatus contains the status of a cluster.
+type ClusterStatus struct {
+	// TODO https://github.com/kubernetes/cluster-registry/issues/28
 }
 
-type URL string
+// ServerAddressByClientCIDR helps the client to determine the server address that they should use, depending on the clientCIDR that they match.
+type ServerAddressByClientCIDR struct {
```
Comment (font) [inline pkg/apis/clusterregistry/types.go]: small nit: can we define this type below the `KubernetesAPIEndpoints` type?

**Comment by perotinus** [comment]
/assign @pmorie 

@pmorie No worries; would it help if I assign things to you directly?

File: pkg/apis/clusterregistry/types.go (hunk 1 above)
Comment (perotinus) [reply pkg/apis/clusterregistry/types.go]: Sure, done.

**Comment by pmorie** [comment]
@perotinus unfortunately I think i need to rejigger my filters to have a hope of making notifications on github useful again.  I actually watch this repo and have every go into my inbox, but it's a flood.

**Comment by pmorie** [comment]
LGTM, i think i may open a PR with a couple tweeks but this is movement in the right direction.

**Comment by madhusudancs** [comment]
/lgtm

Thanks for modifying the endpoint object.

**Comment by perotinus** [comment]
Thanks, everyone!

**Pull request merged**